### PR TITLE
Init/claw

### PR DESCRIPTION
**Pull request by cyrusnaficy**
Draft PR for claw

### CODE REVIEW

File: subsystem/claw.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from robotpy_toolkit_7407.sensors.limit_switches.limit_switch import LimitSwitch
+from robotpy_toolkit_7407.subsystem_templates.drivetrain import SwerveDrivetrain
+import config
+import ctre
+import constants
+import wpilib
+
+class Claw(Subsystem):
+
+    claw_motor: SparkMax = SparkMax(config.claw_motor_extend_id)
```
Comment (mllewellyn) [inline subsystem/claw.py:13]: Same comment as [here](https://github.com/Choate-Robotics/7407-DriveCode-ChargedUp/pull/9#discussion_r1083373806) about the value of initializing dependencies outside of the class that is using them.

File: subsystem/claw.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        # Return through bore encode value 
+        return self.claw_motor.get_sensor_position() / constants.claw_gear_ratio
+
+
+    def zero(self):
+        self.claw_motor.set_sensor_position(0)
+        self.set_angle(0)
+
+    # def set claw motor speed
+    def set_claw_speed(self, speed: float):
```
Comment (mllewellyn) [inline subsystem/claw.py:43]: This is pretty minor, but I would consider using `output` over `speed` here and elsewhere. There's the risk `speed` could be confused with setting the setpoint for a PID based velocity controller, or reading the encoder built into the motor.

File: subsystem/claw.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        self.raw_output = speed
+
+    def open_claw(self):
+        # Set distance forward (closes claw)
+        self.claw_close_piston.set(wpilib.DoubleSolenoid.Value.kForward)
+        self.set_claw_speed(constants.claw_motor_speed)
+        self.is_claw_compressed = True
+
+    def close_claw(self):
+        self.claw_close_piston.set(wpilib.DoubleSolenoid.Value.kReverse)
```
Comment (mllewellyn) [inline subsystem/claw.py:60]: I would consider using the `DoubleSolenoidPiston` wrapper class from the toolkit here because the methods `extend` and `retract` are more readable than `set(wpilib.DoubleSolenoid.Value.kReverse)`.

**Review by mllewellyn — COMMENTED** [review]
Thanks for posting the review! Some comments but overall very good work so far!

File: subsystem/claw.py (hunk 3 above)
Comment (sidharthmrao) [reply subsystem/claw.py:60]: ^ I completely agree with this

**Comment by cyrusnaficy** [comment]
Updated the code according to the comments on my initial commit. Also fixed constants.py merge conflict

**Comment by cyrusnaficy** [comment]
Fixed new conflicting files

File: subsystem/claw.py (hunk 1 above)
Comment (mllewellyn) [reply subsystem/claw.py:14]: Good job adding the params to the constructor. There's a side effect I don't think you realize which requires a bit more change, but it's easy.

TLDR: Remove constructing the spark and piston to robot.py or whichever file you setup this subsystem.

Checkout [this example](https://gist.github.com/mllewellyn/043c61e75dce4d5981fc42d7c21e17e2) I made. See how despite us passing in new objects, the internal ones are still setup (and then thrown away)?

This is undesirable because if there's a side effect to constructing a `SparkMax` (such as a CAN ID being reserved in some system, which I think there is) when multiple tests run one after the other a good test can break because an unrelated test had an unexpected side effect.

Comment (mllewellyn) [reply subsystem/claw.py:14]: Also type hints in the construct would be nice :-) 

Long term we should setup something like mypy to automatically tell us when we need to type hint so that it's not a bother on reviews.

Comment (cyrusnaficy) [reply subsystem/claw.py]: I see what you are trying to say, thanks for the example. I reworked the class so that the motor is passed through the constructor (so no overrides will occur). I also marked its type as a SparkMax so other functions would still show.

Let me know if this is what you were looking for.

File: command/claw.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            self.subsystem.set_angle(math.pi / 2)
+        else:
+            self.subsystem.set_angle(0)
+
+    def execute(self) -> None:
+        pass
+
+    def isFinished(self) -> bool:
+        if self.up:
+            return round(math.degrees(self.subsystem.get_angle())) == 90
```
Comment (mllewellyn) [inline command/claw.py:46]: This doesn't need any changes now, but something that you might encounter while tuning automodes is that this command takes a while to finish because 1 degree is a tight tolerance. If you encounter that you can change to `return abs(value-target) < tolerance`.

File: command/claw.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    def __init__(self, subsystem: Claw, up: bool):
+        super().__init__(subsystem)
+        self.subsystem = subsystem
+        self.up = up
+
+    def initialize(self) -> None:
+        self.subsystem.set_claw_ouput(config.claw_motor_speed)
+        if self.up:
+            self.subsystem.set_angle(math.pi / 2)
```
Comment (mllewellyn) [inline command/claw.py:37]: I think this would be a good place for a constant instead of `math.pt / 2`.

File: command/claw.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+class ClawCrunch(SubsystemCommand[Claw]):
+
+    def __init__(self, subsystem: Claw,  compress: bool):
+        super().__init__(subsystem)
+        self.subsystem = subsystem
+        self.compress = compress
+
+    def initialize(self) -> None:
+        if self.compress:
+            self.subsystem.open_claw()
```
Comment (mllewellyn) [inline command/claw.py:59]: Style question so no wrong answers; Why use a compress parameter in the command, but a separate methods for open and closing the claw on the subsystem?

File: command/drivetrain.py [hunk 7]
```diff
@@ -105,7 +108,11 @@ def execute(self) -> None:
         self.zero()
 
     def isFinished(self) -> bool:
-        return self.zero_success()
+        # self.subsystem.gyro.reset_angle()
```
Comment (mllewellyn) [inline command/drivetrain.py:111]: Where did these changes come from? Are ya'll merging each others unfinished code?

File: subsystem/claw.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        Args:
+            pos (float): Angle in radians (0, 2pi)
+        """
+        self.claw_motor.set_target_position(
+            (pos / (math.pi * 2)) * constants.claw_gear_ratio  # 80 rotations per 360 degrees
+        )   
+
+    def get_angle(self):
+        # Return through bore encode value 
+        return self.claw_motor.get_sensor_position() / constants.claw_gear_ratio
```
Comment (mllewellyn) [inline subsystem/claw.py:42]: What are the unit here? Does this sensor need any scaling?

File: subsystem/claw.py (hunk 1 above)
Comment (mllewellyn) [reply subsystem/claw.py]: Exactly what I was looking for with the spark in the claw subsystem. Don't forget to setup the spark max in robot.py and pass it in though! Or forget about the `SingleSolenoidPiston` :-)

**Review by mllewellyn — COMMENTED** [review]
Nice to see this coming along!

File: command/drivetrain.py (hunk 7 above)
Comment (cyrusnaficy) [reply command/drivetrain.py:111]: Sebastian is merged his elevator code.

File: subsystem/claw.py (hunk 8 above)
Comment (cyrusnaficy) [reply subsystem/claw.py:42]: Radians divided by the gear ratio (gear ratio is a dummy value now as the claw has not been put on the robot yet)

Comment (cyrusnaficy) [reply subsystem/claw.py:42]: I am now passing SparkMax inside of robot_systems. Let me know if that was what you were trying to ask me to implement. I also replaced the piston back to `DoubleSolenoidPiston` as per what @sidharthmrao's told me.

**Comment by cyrusnaficy** [comment]
Tested today on claw. Motor works as intended. Cylinders have yet to be added, but when they are, we will test with those too.
